File: Extras/nengo/pidcontroller.py
'''
Simple altitude-hold PID controller

Copyright (C) 2019 Simon D. Levy

MIT License
'''

class AltitudePidController(object):

    def __init__(self, target, posP, velP, velI, velD, windupMax=10):

        # In a real PID controller, this would be a set-point
        self.target = target

        # Constants
        self.posP = posP
        self.velP = velP
        self.velI = velI
        self.velD = velD
        self.windupMax = windupMax

        # Values modified in-flight
        self.posTarget      = 0
        self.lastError      = 0
        self.integralError  = 0

    def u(self, alt, vel, dt):

        # Compute dzdt setpoint and error
        velTarget = (self.target - alt) * self.posP
        velError = velTarget - vel

        # Update error integral and error derivative
        self.integralError +=  velError * dt
        self.integralError = AltitudePidController._constrainAbs(self.integralError + velError * dt, self.windupMax)
        deltaError = (velError - self.lastError) / dt if abs(self.lastError) > 0 else 0
        self.lastError = velError

        # Compute control u
        return self.velP * velError + self.velD * deltaError + self.velI * self.integralError

    def _constrainAbs(x, lim):

        return -lim if x < -lim else (+lim if x > +lim else x)

File: Extras/servers/haskell/PidController.hs
{--
  PID Controller

  Copyright(C) 2021 Simon D.Levy

  MIT License
--}

module AltitudePidController (AltitudePidController) where

data AltitudePidController = AltitudePidController Double Double Double Double Double Double deriving Show

pid_kp :: AltitudePidController -> Double
pid_kp (AltitudePidController kp _ki _kd _windupMax _lastError _errorIntegral) = kp

pid_ki :: AltitudePidController -> Double
pid_ki (AltitudePidController _kp ki _kd _windupMax _lastError _errorIntegral) = ki

pid_kd :: AltitudePidController -> Double
pid_kd (AltitudePidController _kp _ki kd _windupMax _lastError _errorIntegral) = kd

pid_windup_max :: AltitudePidController -> Double
pid_windup_max (AltitudePidController _kp _ki _kd windupMax _lastError _errorIntegral) = windupMax

makeAltitudePidController :: Double -> Double -> Double -> AltitudePidController
makeAltitudePidController kp ki kd = AltitudePidController kp ki kd 0 0 0
